### PR TITLE
Pass pad_id, to the generate

### PR DESCRIPTION
**Pull request by krammnic**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* #2460

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/eleuther_eval.py [hunk 1]
```diff
@@ -408,12 +408,18 @@ def _model_generate(
             dtype=self._dtype,
             decoder_max_seq_len=self.max_length,
         ):
+            if hasattr(self.tokenizer, "pad_id") and self.tokenizer.pad_id is not None:
+                pad_id = self.tokenizer.pad_id
+            else:
+                pad_id = 0
```
Comment (felipemello1) [inline recipes/eleuther_eval.py:414]: is it `self.tokenizer` or `self._tokenizer`? This also can be in the __init__, since its a constant.

Can you please test it to confirm it works? thank you!

File: recipes/eleuther_eval.py [hunk 2]
```diff
@@ -408,12 +408,18 @@ def _model_generate(
...
+
             toks, _ = generate(
```
Comment (felipemello1) [inline recipes/eleuther_eval.py:416]: do we also need changes in `_model_multimodal_generate`?

Comment (krammnic) [reply recipes/eleuther_eval.py:418]: I think we don't as we don't use `generate` there

**Comment by krammnic** [comment]
It seems to me that both wrappers are not really covered with tests

**Comment by krammnic** [comment]
`TestEleutherEval` has passed locally

**Comment by felipemello1** [comment]
> `TestEleutherEval` has passed locally

can you please run the recipe?

File: recipes/eleuther_eval.py (hunk 2 above)
Comment (felipemello1) [reply recipes/eleuther_eval.py:417]: true, but we are still generating. Does it means that no padding is happening or the arg is just not exposed?

**Comment by krammnic** [comment]
> > `TestEleutherEval` has passed locally
> 
> can you please run the recipe?

It is fine: 

Max memory allocated: 9.61 GB


|    Tasks     |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|--------------|------:|------|-----:|------|---|-----:|---|-----:|
|truthfulqa_mc2|      2|none  |     0|acc   |↑  |0.4175|±  |0.0146|

Furthermore, I've noticed that qwen2_5 eval config is incorrect, so pushed fix for it here.

File: recipes/eleuther_eval.py [hunk 3]
```diff
@@ -316,6 +316,11 @@ def __init__(
         self._dtype = dtype
         self._enable_kv_cache = enable_kv_cache
 
+        if hasattr(self._tokenizer, "pad_id") and self._tokenizer.pad_id is not None:
```
Comment (pbontrager) [inline recipes/eleuther_eval.py:319]: We really need to an an abstract class for our tokenizers, but I think it's safe to assume in the meantime that tokenizer.pad_id exists for each tokenizer as our recipes use this. I think we can skip this hasattr check.

File: recipes/eleuther_eval.py (hunk 2 above)
Comment (pbontrager) [reply recipes/eleuther_eval.py:417]: padded_collate_tiled_images_and_mask used in the _VLMEvalWrapper also takes padding_idx as an input

File: recipes/eleuther_eval.py [hunk 4]
```diff
@@ -361,7 +366,7 @@ def tok_batch_encode(
         x = left_pad_sequence(
             [torch.tensor(x) for x in tokenized_text],
             batch_first=True,
-            padding_value=self._tokenizer.pad_id,
+            padding_value=self.pad_id,
```
Comment (pbontrager) [inline recipes/eleuther_eval.py:369]: I actually think self._tokenizer.pad_id is correct here. I don't think this is the location where the issue is.

File: recipes/eleuther_eval.py (hunk 2 above)
Comment (pbontrager) [reply recipes/eleuther_eval.py:417]: @felipemello1 we don't do batched inference for multimodal

File: recipes/eleuther_eval.py (hunk 4 above)
Comment (krammnic) [reply recipes/eleuther_eval.py:369]: Then it will be inconsistent. As the solution, I assume that we can just not define it in `__init__`, dropping `hasattr` check according to the previous comment, but we still need `is not None` check don't we?

Comment (felipemello1) [reply recipes/eleuther_eval.py:369]: since .pad_id is in every tokenizer, it makes sense to remove it from the __init__, without hasattr, and just doing `self._tokenizer.pad_id` directly. Thanks for asking!

After it I will merge

**Comment by krammnic** [comment]
@felipemello1 lint is fine + integration test passed, so we can start CI

**Review by felipemello1 — APPROVED** [review]
